BaseLib/Functional.h

```diff
@@ -0,0 +1,214 @@
+                     IndexedPlacedPlaceholder<Indices>::value...);
+}
+
+template <int... Indices, typename Object, typename MethodClass,
+          typename ReturnType, typename... Args>
+std::function<ReturnType(Args...)> easyBind_inner(
+    ReturnType (MethodClass::*method)(Args...), Object&& obj,
+    IntegerSequence<Indices...>)
+{
+    return easyBind_innermost<Indices...>(method, std::forward<Object>(obj));
+}
+
+template <int... Indices, typename Object, typename MethodClass,
+          typename ReturnType, typename... Args>
+std::function<ReturnType(Args...)> easyBind_inner(
+    ReturnType (MethodClass::*method)(Args...) const, Object&& obj,
+    IntegerSequence<Indices...>)
+{
+    return easyBind_innermost<Indices...>(method, std::forward<Object>(obj));
+}
+
+/*! Deduces the signature of the call operator of class \c T.
+ *
+ * The matching type of std::function is provided as the member type
+ * \c FunctionType.
+ *
+ * \see http://stackoverflow.com/a/7943765
+ */
+template <typename T>
+struct FunctionTraits
+    : public FunctionTraits<decltype(&std::decay<T>::type::operator())> {
+};
+
+template <typename Object, typename ReturnType, typename... Args>
+struct FunctionTraits<ReturnType (Object::*)(Args...)> {
+    using FunctionType = std::function<ReturnType(Args...)>;
+};
+
+template <typename Object, typename ReturnType, typename... Args>
+struct FunctionTraits<ReturnType (Object::*)(Args...) const> {
+    using FunctionType = std::function<ReturnType(Args...)>;
+};
+
+}  // namespace detail
+
+/*! Convenience wrapper for std::bind().
+ *
+ * This function binds the member function pointer \c member of class \c Object
+ * to the instance \c obj of this class and wraps the result in a std::function
+ * with matching signature.
+ *
+ * The result of this function can be used, e.g., to deduce the signature of the
+ * \c method (which is not possible with the result of std::bind).
+ *
+ * Example:
+ * \code{.cpp}
+ * using std::placeholders;
+ * Object some_object;
+ *
+ * auto f_bind = std::function<ReturnType(Arg1, Arg2, Arg3>(
+ *                  std::bind(&Object::methodWithThreeArguments,
+ *                      std::ref(some_object), _1, _2, _3);
+ *
+ * auto f_easy = easyBind(&Object::methodWithThreeArguments, some_object);
+ * \endcode
+ *
+ * In the example the expressions creating \c f_bind and \c f_easy are
+ * equivalent.
+ *
+ * \note
+ * There is one difference between the behaviour of std::bind and the one of
+ * easyBind: In easyBind \c obj is never copied, instead it will be referenced.
+ * This is in contrast to the behaviour of std::bind, and has been chosen in
+ * order to prevent accidental copies.
+ */
+template <typename Object, typename MethodClass, typename ReturnType,
+          typename... Args>
+typename std::enable_if<
+    std::is_same<MethodClass,
+                 typename std::remove_cv<typename std::remove_pointer<
+                     typename std::decay<Object>::type>::type>::type>::value,
```

Comment:
Since you obviously do some testing: Have you tested with the actual code of this PR? If I remove `std::remove_cv` I get the following error:

```
[ 71%] Building CXX object Tests/CMakeFiles/testrunner.dir/BaseLib/TestFunctional.cpp.o
src/Tests/BaseLib/TestFunctional.cpp: In member function ‘virtual void BaseLib_Functional_Test::TestBody()’:
src/Tests/BaseLib/TestFunctional.cpp:148:59: error: no matching function for call to ‘easyBind(double (A::*)() const, const A*&)’
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                                                           ^
In file included from src/Tests/BaseLib/TestFunctional.cpp:13:0:
src/BaseLib/Functional.h:179:1: note: candidate: template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...), Object&&)
 easyBind(ReturnType (MethodClass::*method)(Args...), Object&& obj)
 ^
src/BaseLib/Functional.h:179:1: note:   template argument deduction/substitution failed:
src/Tests/BaseLib/TestFunctional.cpp:148:59: note:   types ‘ReturnType (MethodClass::)(Args ...)’ and ‘double (A::)() const’ have incompatible cv-qualifiers
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                                                           ^
In file included from src/Tests/BaseLib/TestFunctional.cpp:13:0:
src/BaseLib/Functional.h:194:1: note: candidate: template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...) const, Object&&)
 easyBind(ReturnType (MethodClass::*method)(Args...) const, Object&& obj)
 ^
src/BaseLib/Functional.h:194:1: note:   template argument deduction/substitution failed:
src/BaseLib/Functional.h: In substitution of ‘template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...) const, Object&&) [with Object = const A*&; MethodClass = A; ReturnType = double; Args = {}]’:
src/Tests/BaseLib/TestFunctional.cpp:148:59:   required from here
src/BaseLib/Functional.h:194:1: error: no type named ‘type’ in ‘struct std::enable_if<false, std::function<double()> >’
src/BaseLib/Functional.h:206:55: note: candidate: template<class Object> typename BaseLib::detail::FunctionTraits<T>::FunctionType BaseLib::easyBind(Object&&)
 typename detail::FunctionTraits<Object>::FunctionType easyBind(Object&& obj)
                                                       ^
src/BaseLib/Functional.h:206:55: note:   template argument deduction/substitution failed:
src/Tests/BaseLib/TestFunctional.cpp:148:59: note:   candidate expects 1 argument, 2 provided
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
```